agent/consul/discoverychain/gateway.go

```diff
@@ -128,6 +128,29 @@ func (l *GatewayChainSynthesizer) Synthesize(chains ...*structs.CompiledDiscover
 			return nil, nil, err
 		}
 
+		node := compiled.Nodes[compiled.StartNode]
+		if node.IsRouter() {
+			resolverPrefix := structs.DiscoveryGraphNodeTypeResolver + ":" + node.Name
+
+			// clean out the clusters that will get added for the router
+			for name := range compiled.Nodes {
+				if strings.HasPrefix(name, resolverPrefix) {
+					delete(compiled.Nodes, name)
+				}
+			}
+
+			// clean out the route rules that'll get added for the router
+			filtered := []*structs.DiscoveryRoute{}
+			for _, route := range node.Routes {
+				if strings.HasPrefix(route.NextNode, resolverPrefix) {
+					continue
+				}
+				filtered = append(filtered, route)
+			}
+			node.Routes = filtered
```

Comment:
So what this is doing is dropping any route (specifically the default "catch-all" mentioned in https://developer.hashicorp.com/consul/docs/connect/config-entries/service-router#path-prefix-matching) that would route to the problematic generated resolver?

In "clean out the clusters" above, is what's happening removing the generated resolver node (with it's corresponding compiled clusters) from the discovery chain?

Reply:
Yeah, so this is removing a route that goes to a resolver matching the name of the router -- which is the catch-all route.

The other step is removing the same resolver from the list of nodes in the full discovery chain, which is what is used to generate the cluster.

The node removal is necessary to keep this from blocking configuration push down, the route removal is to keep envoy from trying to resolve some random hostname and having a bogus routing rule.